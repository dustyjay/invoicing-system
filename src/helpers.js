export const CONSTANTS = {
  INVOICES: 'invoices',
  SHOW_PREVIEW: 'show-preview',
  USER_DATA: 'user-data',
  INVOICE_NUMBER_LENGTH: 5
};

export const formatInvoiceNumber = number => {
  const num = Math.ceil(number);
  const { length } = `${num}`;
  let zeroDigits = CONSTANTS.INVOICE_NUMBER_LENGTH - length;
  let finalNumber = '';
  while (zeroDigits > 0) {
    finalNumber += '0';
    zeroDigits -= 1;
  }
  return finalNumber + num;
};

export const formatDate = date => {
  if (date) return new Date(date).toDateString();
  return '';
};

export const formatCurrency = value => {
  const formatter = new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD'
  });

  return formatter.format(value);
};

export const setStorageItem = (key, value) => {
  localStorage.setItem(key, JSON.stringify(value));
};

export const getStorageItem = key => {
  const item = localStorage.getItem(key);
  if (item) return JSON.parse(item);
};

export const removeStorageItem = key => {
  localStorage.removeItem(key);
};
