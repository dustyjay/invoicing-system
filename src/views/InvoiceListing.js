import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import InvoiceCard from '../shared/InvoiceCard';
import { CONSTANTS, getStorageItem, setStorageItem } from '../shared/invoices';
import AddIcon from '../assets/images/icons/plus-white.png';
import SearchIcon from '../assets/images/icons/search-icon.png';
import UserIcon from '../assets/images/icons/user.svg';
import AddPersonalData from '../shared/AddPersonalData';

const InvoiceListing = () => {
  const [invoices, setInvoices] = useState([]);
  const [showUserData, setShowUserData] = useState(false);
  const history = useHistory();

  useEffect(() => {
    const allInvoices = getStorageItem(CONSTANTS.INVOICES) || [];
    setInvoices(allInvoices);
  }, []);

  const goToInvoice = invoiceId => {
    history.push(`/${invoiceId}`);
  };

  const handleOnDelete = index => {
    const allInvoices = [...invoices];
    allInvoices.splice(index, 1);
    setInvoices(allInvoices);
    setStorageItem(CONSTANTS.INVOICES, invoices);
  };

  const handleOnView = index => {
    setStorageItem(CONSTANTS.SHOW_PREVIEW, true);
    goToInvoice(index);
  };

  const addNewInvoice = () => {
    setStorageItem(CONSTANTS.INVOICES, [...invoices, getDefaultInvoice()]);
    goToInvoice(invoices.length + 1);
  };

  const getDefaultInvoice = () => {
    const dateIssued = new Date().toISOString().split('T')[0];
    return {
      clientName: '',
      dateIssued,
      dueDate: '',
      discount: '',
      tax: '',
      items: [],
      userData: getStorageItem(CONSTANTS.USER_DATA) || {}
    };
  };

  return (
    <section>
      <AddPersonalData
        show={showUserData}
        closeModal={() => setShowUserData(false)}
      />
      <div className="invoice__title">
        <h1>üìù Invoices</h1>
        <div className="d-flex">
          <button
            className="fab fab--back user show-mobile"
            onClick={() => setShowUserData(true)}
          >
            <img src={UserIcon} alt="view your data" />
          </button>
          {invoices.length > 0 && (
            <div className="fab-box">
              <button className="fab fab--add" onClick={addNewInvoice}>
                <img src={AddIcon} alt="Add new invoice" />
                <span className="show-mobile">Add New</span>
              </button>
            </div>
          )}
        </div>
      </div>
      {invoices.length ? (
        <div className="invoice__list">
          {invoices.map((el, index) => (
            <InvoiceCard
              {...el}
              key={index}
              onEdit={() => goToInvoice(index + 1)}
              onView={() => handleOnView(index + 1)}
              onDelete={() => handleOnDelete(index)}
            />
          ))}
        </div>
      ) : (
        <div className="invoice-item__empty">
          <i>
            <img src={SearchIcon} alt="Empty invoice list" />
          </i>
          <p>No invoices yet!</p>
          <button className="fab fab--add" onClick={addNewInvoice}>
            <span>Add New Invoice</span>
          </button>
        </div>
      )}
    </section>
  );
};

export default InvoiceListing;
