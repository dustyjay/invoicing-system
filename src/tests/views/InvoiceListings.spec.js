import React from 'react';
import { render, cleanup } from '@testing-library/react';
import InvoiceListing from '../../views/InvoiceListing';
import { CONSTANTS, setStorageItem } from '../../helpers';
import { INVOICES } from '../testData';

describe('Invoice Listing', () => {
  let wrapper;
  let container;

  beforeEach(() => {
    wrapper = render(<InvoiceListing />);
    container = wrapper.container;
  });

  afterEach(() => {
    wrapper.unmount();
    cleanup();
  });

  test('Invoice matches snapshot', () => {
    const tree = wrapper.container;
    expect(tree).toMatchSnapshot();
  });

  test('has button to view user data', () => {
    const button = container.querySelector('button.user');
    const userImage = button.querySelector('img[alt="view user data"]');
    expect(button).toBeTruthy();
    expect(userImage).toBeTruthy();
  });

  test('renders empty state if there are no invoices', () => {
    const emptyContainer = container.querySelector('.invoice-item__empty');
    const invoiceList = container.querySelector('.invoice__list');
    expect(emptyContainer).toBeTruthy();
    expect(invoiceList).toBeFalsy();
  });

  test('renders number of available invoices in storage', async () => {
    setStorageItem(CONSTANTS.INVOICES, INVOICES);
    const { container } = render(<InvoiceListing />);
    const emptyContainer = container.querySelector('.invoice-item__empty');
    const invoiceList = container.querySelectorAll('.invoice-card');
    expect(invoiceList.length).toEqual(INVOICES.length);
    expect(emptyContainer).toBeFalsy();
  });
});
