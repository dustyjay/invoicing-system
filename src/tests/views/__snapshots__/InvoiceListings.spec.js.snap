// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoice Listing Invoice matches snapshot 1`] = `
<div>
  <section>
    <div
      class="invoice__title"
    >
      <h1>
        üìù Invoices
      </h1>
      <div
        class="d-flex"
      >
        <button
          class="fab fab--back user show-mobile"
        >
          <img
            alt="view user data"
            src="user.svg"
          />
        </button>
      </div>
    </div>
    <div
      class="invoice-item__empty"
    >
      <i>
        <img
          alt="Empty invoice list"
          src="search-icon.png"
        />
      </i>
      <p>
        No invoices yet!
      </p>
      <button
        class="fab fab--add"
      >
        <span>
          Add New Invoice
        </span>
      </button>
    </div>
  </section>
</div>
`;
